<a href="https://colab.research.google.com/github/helenanaa/Global-Earth-Temperatures/blob/main/Checkpoint5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **GLOBAL EARTH TEMPERATURES**

#**Import Library**

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from prettytable import PrettyTable
import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier # Algoritma C4.5
from sklearn.preprocessing import StandardScaler
from sklearn import preprocessing
from sklearn import metrics

In [ ]:
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.cluster import KMeans
from sklearn.cluster import AgglomerativeClustering
from sklearn import metrics
from sklearn.metrics import accuracy_score

#**BUSINESS UNDERSTANDING**

##**Business Objective**

Tujuan bisnis dari dataset ini adalah memantau dan menganalisis iklim, mengembangan solusi untuk mitiasi dan adaptasi perubahan iklim, meningkatkan kesdaran dan pemahaman tentang perubahan iklim, dan memperkuat kerjasama internasional.

##**Assess Situation**

Dataset ini merupakan sumber informasi yang berharga bagi berbagai pihak yang ingin memahami, memantau, dan mengatasi perubahan iklim.

##**Data Mining Goals**

Tujuan dari data mining pada dataset ini adalah untuk mendapatkan wawasan yang berharga tentang pola dan tren dalam perubahan suhu global serta faktor-faktor yang mempengaruhinya.

##**Project Plan**

Rencana dari proyek untuk menganalisis dataset ini dimulai dari pengumpulan/menggabungkan data dari berbagai sumber yang ada. selanjutnya pemrosesan data untuk memeriksa kualitas data. pengembangan alat untuk visualisasi dan analisis data bumi dan membangun antarmuka pengguna yang intutif untuk mengakses data. selain itu, memperbarui dataset secara berkala dengan data baru dan memantau kualitas data atau melakukan perbaikan jika dibutuhkan.

***Checkpoint 2***

#**DATA UNDERSTANDING**

##**Collect Initial Data**

dataset diambil dari kaggle, di download lalu di extract dan upload pada google colab. https://www.kaggle.com/datasets/joebeachcapital/global-earth-temperatures

In [ ]:
df = pd.read_csv("Global Temperatures.csv")
df.head()

Year   Month   Month Anomaly  Month Unc.   Annual Anomaly  Annual Unc.  \
0   1850       1          -0.801       0.482              NaN          NaN   
1   1850       2          -0.102       0.592              NaN          NaN   
2   1850       3          -0.119       0.819              NaN          NaN   
3   1850       4          -0.485       0.575              NaN          NaN   
4   1850       5          -0.351       0.549              NaN          NaN   

  Five-year Anomaly  Five-year Unc.   Ten-year Anomaly   Ten-year Unc.  \
0               NaN             NaN                NaN             NaN   
1               NaN             NaN                NaN             NaN   
2               NaN             NaN                NaN             NaN   
3               NaN             NaN                NaN             NaN   
4               NaN             NaN                NaN             NaN   

   Twenty-year Anomaly   Twenty-year Unc.  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
3                  NaN                NaN  
4                  NaN                NaN

##**Describe Data**

####Informasi Dasar

pada dataset ini memiliki 2077 baris dan 12 kolom

In [ ]:
df.shape

(2077, 12)

####Informasi Lanjutan

In [ ]:
print:(df)

dataset ini memiliki 12 kolom yang dimana ada: **Year**: Kolom ini berisi tahun pengukuran.
**Month**: Kolom ini berisi bulan pengukuran.
**Monthly Anomaly**: Kolom ini berisi anomali suhu bulanan. Anomali suhu adalah perbedaan antara suhu bulanan rata-rata dan rata-rata suhu bulanan untuk periode tertentu dalam sejarah.
**Monthly Unc**: Kolom ini berisi ketidakpastian (uncertainty) untuk anomali suhu bulanan.
**Annual Anomaly**: Kolom ini berisi anomali suhu tahunan. Anomali suhu tahunan adalah perbedaan antara suhu tahunan rata-rata dan rata-rata suhu tahunan untuk periode tertentu dalam sejarah.
**Annual Unc**: Kolom ini berisi ketidakpastian (uncertainty) untuk anomali suhu tahunan.
**Five-Year Anomaly**: Kolom ini berisi anomali suhu lima tahunan. Anomali suhu lima tahunan adalah perbedaan antara suhu rata-rata lima tahunan dan rata-rata suhu lima tahunan untuk periode tertentu dalam sejarah.
**Five-Year Unc**: Kolom ini berisi ketidakpastian (uncertainty) untuk anomali suhu lima tahunan.
**Ten-Year Anumaly**: Kolom ini berisi anomali suhu sepuluh tahunan. Anomali suhu sepuluh tahunan adalah perbedaan antara suhu rata-rata sepuluh tahunan dan rata-rata suhu sepuluh tahunan untuk periode tertentu dalam sejarah.
**Ten-Year Unc**: Kolom ini berisi ketidakpastian (uncertainty) untuk anomali suhu sepuluh tahunan.
**Twenty-Year Anomaly**: Kolom ini berisi anomali suhu dua puluh tahunan. Anomali suhu dua puluh tahunan adalah perbedaan antara suhu rata-rata dua puluh tahunan dan rata-rata suhu dua puluh tahunan untuk periode tertentu dalam sejarah.
**Twenty-Year Unc**: Kolom ini berisi ketidakpastian (uncertainty) untuk anomali suhu dua puluh tahunan.

####Informasi Statistik Deskriptif

In [ ]:
df.describe()

Year        Month   Month Anomaly
count  2077.000000  2077.000000     2077.000000
mean   1936.041887     6.497352        0.039636
std      49.976436     3.454161        0.418716
min    1850.000000     1.000000       -0.970000
25%    1893.000000     3.000000       -0.253000
50%    1936.000000     6.000000       -0.039000
75%    1979.000000     9.000000        0.252000
max    2023.000000    12.000000        1.499000

**Count**: jumlah baris yang ada dalam dataset pada kolom Year,Month: dan Monthly Anomaly, jumlah total baris adalah 2077
**Mean**: rata-rata nilai dalam kolom. rata-rata tahun adalah  sekitar 1936.041887, rata-rata bulan sekitar 3.454161, dan rata-rata anomali bulanan adalah sekitar 0.039636.
**Std**: standar deviasi dari nilai kolom. dapat mengukur seberapa jauh nilai yang tersebar dari rata-rata. standar deviasi tahun adalah sekitar 49.98, standar deviasi bulan adalah sekitar 3.45, dan standar deviasi anomali bulanan adalah sekitar 0.419.
**Min**: nilai minimum dalam kolom. misalnya nilai minimum tahun adalah 1850, nilai minimum bulan adalah 1, dan nilai minimum anomali ulanan adalah -0.97
**25%,50%,75%**: kuarti pertama 25%, median 50%, dan kuartil ketiga 75%. kuartil ini adalah nilai yang membagi data menjadi empat bagian yang sama besar.
**Max**: Nilai maksimum dalam kolom. Misalnya, nilai maksimum tahun adalah 2023, nilai maksimum bulan adalah 12, dan nilai maksimum anomali bulanan adalah 1.499.

##**Exploratory Data Analysis**

Setelah visualisasi ditampilkan, wajib diberikan penjelasan yaa dari visualisasi tersebut (Lebih baik tambahkan asumsi seperti, kolom a berpengaruh negatif terhadap kolom b, apabila nilai kolom a meningkat, maka nilai kolom b menurun)

Untuk sel nya silahkan ditambahkan sendiri menyesuaikan dengan seberapa banyak visualisasi yang kalian buat

TAMBAHKAN VISUALISASI KORELASI BESERTA PENJELASANNYA LEBIH BAIK

##**Verify Data Quality**

####Missing Values

In [ ]:
df.isnull().any()

Year                     False
 Month                   False
Monthly Anomaly          False
 Monthly Unc.            False
  Annual Anomaly         False
 Annual Unc              False
Five-Year Anomaly        False
 Five-Year Unc.          False
 Ten-Year Anomaly        False
 Ten-Year Unc.           False
  Twenty-Year Anomaly    False
 Twenty-year Unc.        False
dtype: bool

nilai False, berarti tidak ada nilai nol (nilai kosong) dalam kolom tersebut, sementara jika nilainya adalah True, menunjukkan bahwa ada setidaknya satu nilai nol (nilai kosong) dalam kolom. yang berarti dataset ini bernilai False

####Outliers Values

In [ ]:
print(outliers)

[]


dataset ini tidak memiliki outliers

####Duplicated Values

In [ ]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
2072    False
2073    False
2074    False
2075    False
2076    False
Length: 2077, dtype: bool

nilai duplikat bersifat False

####Inconsistent Values / Noise (Optional)

In [ ]:
df.dtypes

Year                       int64
 Month                     int64
Monthly Anomaly          float64
 Monthly Unc.             object
  Annual Anomaly          object
 Annual Unc               object
Five-Year Anomaly         object
 Five-Year Unc.           object
 Ten-Year Anomaly         object
 Ten-Year Unc.            object
  Twenty-Year Anomaly     object
 Twenty-year Unc.         object
dtype: object

***Checkpoint 3***

In [ ]:
print((df.isna().sum() / len(df)) * 100)

 Year                   0.0
 Month                  0.0
 Month Anomaly          0.0
 Month Unc.             0.0
  Annual Anomaly        0.0
 Annual Unc.            0.0
Five-year Anomaly       0.0
 Five-year Unc.         0.0
  Ten-year Anomaly      0.0
  Ten-year Unc.         0.0
 Twenty-year Anomaly    0.0
  Twenty-year Unc.      0.0
dtype: float64


hasil diatas merupakan hasil dari operasi yang dilakukan pada DataFrame atau series pandas. yang berarti tidak ada nilai yang kosong (NaN) dalam setiap kolom DataFrame. dan dapat dikatakan bahwa dataset ini tidak memiliki nilai yang hilang/kosong.

In [ ]:
df = pd.read_csv("Global Temperatures.csv")
df.head()

Year   Month   Month Anomaly  Month Unc.   Annual Anomaly  Annual Unc.  \
0   1850       1          -0.801       0.482              NaN          NaN   
1   1850       2          -0.102       0.592              NaN          NaN   
2   1850       3          -0.119       0.819              NaN          NaN   
3   1850       4          -0.485       0.575              NaN          NaN   
4   1850       5          -0.351       0.549              NaN          NaN   

  Five-year Anomaly  Five-year Unc.   Ten-year Anomaly   Ten-year Unc.  \
0               NaN             NaN                NaN             NaN   
1               NaN             NaN                NaN             NaN   
2               NaN             NaN                NaN             NaN   
3               NaN             NaN                NaN             NaN   
4               NaN             NaN                NaN             NaN   

   Twenty-year Anomaly   Twenty-year Unc.  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
3                  NaN                NaN  
4                  NaN                NaN

In [ ]:
pd.DataFrame(df.isna().sum() / len(df) * 100, columns=['Null Ratio %'])

Null Ratio %
 Year                          0.0
 Month                         0.0
 Month Anomaly                 0.0
 Month Unc.                    0.0
  Annual Anomaly               0.0
 Annual Unc.                   0.0
Five-year Anomaly              0.0
 Five-year Unc.                0.0
  Ten-year Anomaly             0.0
  Ten-year Unc.                0.0
 Twenty-year Anomaly           0.0
  Twenty-year Unc.             0.0

pada data ini memberikan gambaran tentang keadaan data, dengan menunjukkan bahwa tidak ada nilai kosong dalam setiap kolom DataFrame, atau persentase nilai kosong sangat rendah (nol).

In [ ]:
results = []

cols = df.select_dtypes(include=['float64', 'int64'])

for col in cols:
  q1 = df[col].quantile(0.25)
  q3 = df[col].quantile(0.75)
  iqr = q3 - q1
  lower_bound = q1 - 1.5*iqr
  upper_bound = q3 + 1.5*iqr
  outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
  percent_outliers = (len(outliers)/len(df))*100
  results.append({'Kolom': col, 'Persentase Outliers': percent_outliers})

results_df = pd.DataFrame(results)
results_df.set_index('Kolom', inplace=True)
results_df = results_df.rename_axis(None, axis=0).rename_axis('Kolom', axis=1)

display(results_df)

Kolom           Persentase Outliers
 Year                      0.000000
 Month                     0.000000
 Month Anomaly             2.070294

hasil data diatas memberikan informasi tentang persentase outliers dalam setiap kolom DataFrame. Outliers adalah nilai yang secara signifikan berbeda dari sebagian besar nilai dalam kumpulan data, dan perhitungan persentase outliers dapat memberikan pemahaman tentang seberapa jauh distribusi data dari nilai-nilai yang dianggap normal. pada persentase 0.000000 menunjukkan bahwa tidak ada outliers dalam kolom tersebut, sementara persentase 2.070294 menunjukkan bahwa sekitar 2.07% dari nilai dalam kolom 'Month Anomaly' dianggap sebagai outliers.

In [ ]:
df_corr = df.corr(numeric_only=True)
px.imshow(df_corr)

pada data ini berfungsi untuk menghitung korelasi hanya untuk kolom-kolom dengan tipe data numerik, setelah menghitung korelasi, hasilnya, yang berupa DataFrame korelasi, kemudian dapat digunakan untuk visualisasi.

In [ ]:
dtc = DecisionTreeClassifier()

In [ ]:
print((df.isna().sum() / len(df)) * 100)

 Year                       0.0
 Month                      0.0
 Month Anomaly              0.0
 Month Unc._  NaN           0.0
 Month Unc._0.029           0.0
                           ... 
  Twenty-year Unc._0.107    0.0
  Twenty-year Unc._0.108    0.0
  Twenty-year Unc._0.11     0.0
  Twenty-year Unc._0.111    0.0
  Twenty-year Unc._0.112    0.0
Length: 4305, dtype: float64


pada data ini memberikan gambaran tentang seberapa banyak nilai nol (atau null) yang ada dalam setiap kolom DataFrame. data ini dapat memberitahu seberapa banyak data yang hilang atau tidak tersedia dalam dataset.

***Check Point 5***

In [ ]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV

In [ ]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier()

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import SelectKBest, chi2

In [ ]:
data = pd.read_csv('Global Temperatures.csv')

In [ ]:
print(data.columns)

Index([' Year', ' Month', ' Month Anomaly', ' Month Unc.', '  Annual Anomaly',
       ' Annual Unc.', 'Five-year Anomaly', ' Five-year Unc.',
       '  Ten-year Anomaly', '  Ten-year Unc.', ' Twenty-year Anomaly',
       '  Twenty-year Unc.'],
      dtype='object')


In [ ]:
print([col for col in df.columns if 'Month' in col])

[' Month', ' Month Anomaly', ' Month Unc.']


In [ ]:
df.head()

Year   Month   Month Anomaly  Month Unc.   Annual Anomaly  Annual Unc.  \
0   1850       1          -0.801       0.482              NaN          NaN   
1   1850       2          -0.102       0.592              NaN          NaN   
2   1850       3          -0.119       0.819              NaN          NaN   
3   1850       4          -0.485       0.575              NaN          NaN   
4   1850       5          -0.351       0.549              NaN          NaN   

  Five-year Anomaly  Five-year Unc.   Ten-year Anomaly   Ten-year Unc.  \
0               NaN             NaN                NaN             NaN   
1               NaN             NaN                NaN             NaN   
2               NaN             NaN                NaN             NaN   
3               NaN             NaN                NaN             NaN   
4               NaN             NaN                NaN             NaN   

   Twenty-year Anomaly   Twenty-year Unc.  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
3                  NaN                NaN  
4                  NaN                NaN

In [ ]:
x = df.drop('Month', axis=1)
y = df['Month']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [ ]:
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10]
}

variabel ini merupakan  sebuah dictionary yang digunakan dalam proses tuning hyperparameter dengan metode grid search atau random search. Dictionary ini berisi daftar hyperparameter yang ingin diuji bersama dengan kumpulan nilai yang mungkin untuk setiap hyperparameter.

In [ ]:
# Inisialisasi model clustering (misalnya KMeans)
kmeans = KMeans(n_clusters=4)
kmeans.fit(df)

# Inisialisasi selector fitur
selector = SelectKBest(score_func=chi2, k='all')

# Menggunakan metode fit_transform pada selector
x_selected = selector.fit_transform(df, kmeans.labels_)

# Mengambil skor fitur
feature_ranks = selector.scores_

# Mengambil nama fitur
feature_names = df.columns

# Membuat DataFrame untuk menyimpan skor fitur
feature_ranks_df = pd.DataFrame({'Feature': feature_names, 'Rank': feature_ranks})

# Mengurutkan DataFrame berdasarkan skor fitur
feature_ranks_df = feature_ranks_df.sort_values(by='Rank', ascending=False)

# Membuat plot
plt.figure(figsize=(10, 6))
plt.barh(feature_ranks_df['Feature'], feature_ranks_df['Rank'], color='skyblue')
plt.xlabel('Rank')
plt.ylabel('Feature')
plt.title('Feature Ranking K-Means')
plt.gca().invert_yaxis()
plt.show()

variable diatas merupakan bagaimana melakukan analisis fitur menggunakan KMeans clustering dan SelectKBest untuk memilih fitur-fitur terbaik berdasarkan skor chi-square.

In [ ]:
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score
import numpy as np

In [ ]:
metrics = []
scores_kmeans = []
scores_hc = []

dunn_score_kmeans = dunn_index(df, kmeans.labels_)
dunn_score_hc = dunn_index(df, hierarchy_clust)

metrics.append("Dunn Index")
scores_kmeans.append(dunn_score_kmeans)
scores_hc.append(dunn_score_hc)

df_scores = pd.DataFrame({'': metrics,
                          'KMeans Score': scores_kmeans,
                          'Agglomerative Clustering Score': scores_hc
})

df_scores = df_scores.set_index('')
df_scores = df_scores.T

display(df_scores)

variabel diatas menghitung dan menyimpan skor Dunn Index dari dua metode clustering yang berbeda, yaitu KMeans dan Agglomerative Clustering lalu skor-skor tersebut dimasukkan ke dalam sebuah DataFrame untuk keperluan pembandingan.

###**K-Means**###

In [ ]:
param_grid = {
    'n_clusters': [3, 4, 5, 6, 7, 8, 9, 10],
    'init': ['k-means++', 'random'],
    'algorithm': ['lloyd', 'elkan']
}

# Evaluasi kualitas klastering menggunakan Silhouette Score
best_score = -1
best_params = None

for n_clusters in param_grid['n_clusters']:
    for init in param_grid['init']:
        for algorithm in param_grid['algorithm']:
            kmeans = KMeans(n_clusters=n_clusters, init=init, algorithm=algorithm, random_state=42, n_init='auto')
            labels = kmeans.fit_predict(df)
            silhouette_avg = silhouette_score(df, labels)
            if silhouette_avg > best_score:
                best_score = silhouette_avg
                best_params = {'n_clusters': n_clusters, 'init': init, 'algorithm': algorithm}

# Menampilkan parameter dengan Silhouette Score terbaik
print("\nParameter terbaik :")
print(best_params)
print("Silhouette Score :", best_score)

variabel diatas merupakan implementasi grid search untuk menemukan parameter terbaik untuk model KMeans clustering berdasarkan evaluasi menggunakan Silhouette Score.

In [ ]:
hierarchical_cluster_tuned = AgglomerativeClustering(metric='cosine', linkage='single')

hierarchical_cluster_tuned

AgglomerativeClustering(linkage='single', metric='cosine')

Agglomerative Clustering adalah sebuah algoritma clustering hierarkis yang memulai dengan setiap titik data sebagai klaster terpisah, lalu secara iteratif menggabungkan dua klaster yang paling dekat satu sama lain berdasarkan kriteria tertentu, seperti jarak Euclidean atau kesamaan kosinus, sampai semua titik data tergabung dalam satu klaster atau mencapai jumlah klaster yang diinginkan.